department-of-veterans-affairs/caseflow
# TYLERB/APPEALS-13118: Team Management request submission behavior

Resolves [APPEALS-13118](https://vajira.max.gov/browse/APPEALS-13118)
Also Resolves: [APPEALS-13123](https://vajira.max.gov/browse/APPEALS-13123)
Also Resolves: [APPEALS-13124](https://vajira.max.gov/browse/APPEALS-13124)

### Description
Added form submission behavior to the Vha Help page team management form. This includes the success message, the creation of the membership request objects, the sending of the emails to the user who submits the form, and the emails to the organizations that the user has requested to join.  

Link to the designs: https://bipuiux.invisionapp.com/console/share/Z49BG5ERNDT/980289240

### Acceptance Criteria
- [ ] Success Banner
  - [ ] Success Banner header matches design text
  - [ ] Success Banner body text matches designs
  - [ ] Success Banner body successfully reflects which orgs the user requested to join
- [ ] Form State
  - [ ] The form state after submission is properly updated to reflect all the new membership requests
  - [ ] The form state is cleared after submission
- [ ] The form submission successfully generates Membership Request to all the selected orgs for the user that submitted the form
- [ ] The membership requests contain a note if the user added one when filling out the form.
- [ ] Emails
  - [ ] User email 
    - [ ] There is an email sent to the user that matches the designs
    - [ ] The user email updates to reflect all of the requests that were made to the various orgs.
  - [ ] Admin Emails
    - [ ] There should be an email sent to each org that has a new access request by the user that matches the designs
    - [ ] The to email address should be correct for all the orgs
    - [ ] The Request to Member Services – Health Eligibility Center should have two emails to two different email addresses  

### Testing Plan
1. Data setup. You will have to either create a new user for each time you run through the form or use one and continuously delete all of it's membership requests
```ruby
#create a new user
new_user = FactoryBot.create(:user, css_id: "NEW USER 1", email: "test@test.com", full_name: "Gaius Baelsar")
new_user.save

# Enable or disable feature toggle as needed
FeatureToggle.enable!(:program_office_team_management)
#FeatureToggle.disable!(:program_office_team_management)
```
Also we need a way to view the emails that are sent. Following the instructions here will let us do that.
https://github.com/department-of-veterans-affairs/caseflow/wiki/Testing-Action-Mailer-Messages-Locally
You will have also run your backend with the environmental variable so it actually sends the emails instead of throwing them away.
You will also have to start the jobs process in another terminal so that the email job will actually run. Hopefully after all this setup, the emails should display properly.
```
WITH_TEST_EMAIL_SERVER=true make run-low-priority
```
2. Login as your new user. E.g. NEW USER 1
3. Navigate to the vha help page at the url /vha/help
4. Select the VHA checkbox, enter a note into the textbox, and submit the form
5. Verify that the success message matches the designs
![image](https://user-images.githubusercontent.com/109369527/221704425-7ba10a2c-148b-444a-a68b-5a7d1c7eb0e4.png)

6. Verify that the form was reset
7. Verify that the VHA checkbox is disabled and the disabled checkbox notice is visible
![image](https://user-images.githubusercontent.com/109369527/221704661-8ef921a8-6d9e-4de4-b1e1-c9cb67b08414.png)
8. Verify that the membership object to the VHA org was created and contains the note that was entered in the textbox
```ruby
my_user = User.find_by(css_id: "NEW USER 1")
my_user.membership_requests
```
![image](https://user-images.githubusercontent.com/109369527/221705192-3c3511a5-0d52-4ed9-95fc-cf07a75e379b.png)

9. Verify that the user email was sent and you can view it at http://localhost:1080/
![image](https://user-images.githubusercontent.com/109369527/221705686-313c5602-2b12-4276-9bde-df908a024927.png)

10. Verify that the VHA admin email was sent and you can view it at http://localhost:1080/
![image](https://user-images.githubusercontent.com/109369527/221705855-81872f87-c707-43e9-ba7e-c422b2a48576.png)

11. Create a new user or delete that users membership requests and reload the page
```ruby
my_user.membership_requests.each(&:delete)
my_user.save
my_user.reload
```
12. Select the Member Services – Health Eligibility Center checkbox and submit the form
![image](https://user-images.githubusercontent.com/109369527/221707059-8ba8b58c-e1ef-417c-bb95-adf22278f265.png)

13. Verify the form reset, the checkbox disable, and the disabled checkbox notice again.
![image](https://user-images.githubusercontent.com/109369527/221707428-c856f33d-d585-42c0-ac33-c92c4aa37b60.png)

14. Verify the success message matches the designs
15. Verify that the user email was sent and you can view it at http://localhost:1080/
16. Verify that two Member Services – Health Eligibility Center admin emails were sent and you can view both at http://localhost:1080/
![image](https://user-images.githubusercontent.com/109369527/221707756-7609f624-ae40-4254-945a-4c71b2741ac3.png)
![image](https://user-images.githubusercontent.com/109369527/221707860-82d63f9e-b9a0-4021-ac1d-dccbfd9040cc.png)

17. Create a new user or delete that user's membership requests and reload the page
18. Try any additional combination of checkboxes on the form and verify the functionality.

### Storybook Story
* [ ] Updated the VhaHelp storybook file
* [ ] Updated the VhaMembershipRequestsForm storybook file
